src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateClonedParameterSymbolForBeginAndEndInvoke.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class SourceDelegateClonedParameterSymbolForBeginAndEndInvoke : SourceClonedParameterSymbol
+    {
+        internal SourceDelegateClonedParameterSymbolForBeginAndEndInvoke(SourceParameterSymbol originalParam, Symbol newOwner, int newOrdinal)
+            : base(originalParam, newOwner, newOrdinal, suppressOptional: true)
+        {
+        }
+
+        internal override bool IsCallerFilePath => _originalParam.IsCallerFilePath;
+
+        internal override bool IsCallerLineNumber => _originalParam.IsCallerLineNumber;
+
+        internal override bool IsCallerMemberName => _originalParam.IsCallerMemberName;
+
+        internal override int CallerArgumentExpressionParameterIndex => throw ExceptionUtilities.Unreachable;
```

Comment:
@AlekseyTs The created parameter may not be optional as you stated in the other PR. However, I think this is still unreachable. We attempt to bind default arguments only if overload resolution succeeds.

If overload resolution succeeds, then an explicit value was passed to the parameter, meaning we are not going to access this property.

Reply:
> If overload resolution succeeds, then an explicit value was passed to the parameter, meaning we are not going to access this property.

That maybe true for EndEnvoke (I assume because parameters are ref/out/in), but why this is true for BeginInvoke is not obvious to me.

> I'm open to adding any test cases you'd like if the existing tests are not enough

I think we should have targeted test for Begin/EndInvoke invocations for scenarios when cloned parameters are optional and have the attribute on them.

Reply:
Both `BeginInvoke` and `EndInvoke` has extra parameter(s) at the end (`AsyncCallBack` and `object` for `BeginInvoke`, and `IAsyncResult` for `EndInvoke`). If overload resolution succeeds, then these parameters must have been supplied, meaning all the previous parameters has explicitly been passed.

I'll add some tests.

Reply:
> Both `BeginInvoke` and `EndInvoke` has extra parameter(s) at the end (`AsyncCallBack` and `object` for `BeginInvoke`, and `IAsyncResult` for `EndInvoke`). If overload resolution succeeds, then these parameters must have been supplied, meaning all the previous parameters has explicitly been passed.

I see now why this is wrong.